src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp

```diff
@@ -0,0 +1,390 @@
+        rich_input_spectrum[it->second].setMetaValue("IonName", rich_theoretical_spec[it->first].getMetaValue("IonName"));
+    }
+    Param sap = sa.getParameters();
+    rich_input_spectrum.setMetaValue("fragment_match_tolerance", sap.getValue("tolerance"));
+    return rich_input_spectrum;
+  }
+
+  void SpectrumAnnotator::addIonMatchStatistics(PeptideIdentification& pi, const MSSpectrum<Peak1D>& spec, const TheoreticalSpectrumGenerator& tg, const SpectrumAlignment& sa) const
+  {
+    if (!spec.empty())
+    {
+      for (vector<PeptideHit>::iterator ph = pi.getHits().begin(); ph != pi.getHits().end(); ++ph)
+      {
+        RichPeakSpectrum rich_spec = annotateMatches(*ph, spec, tg, sa);
+        rich_spec.sortByIntensity();
+
+        StringList ions;
+        double sum_intensity = 0;
+        double match_intensity = 0;
+        vector<double> fragmenterrors, intensities, mzs; // sorted by ascending intensity via rich_spec.sortByIntensity for topN statistics
+        fragmenterrors.reserve(rich_spec.size());
+        intensities.reserve(rich_spec.size());
+        mzs.reserve(rich_spec.size());
+
+        double nint = 0;
+        double cint = 0;
+
+        StringList allowed_types = ListUtils::create<String>("y,b,a,c,x,z");
+        map<String, vector<bool> > ion_series;
+        for (StringList::iterator st = allowed_types.begin(); st != allowed_types.end(); ++st)
+        {
+          ion_series.insert(make_pair(*st, vector<bool>(ph->getSequence().size()-1, false)));
+        }
+
+        for (RichPeakSpectrum::Iterator it = rich_spec.begin(); it != rich_spec.end(); ++it)
+        {
+          sum_intensity += it->getIntensity();
+          if (it->metaValueExists("IonMatchError"))
+          {
+            fragmenterrors.push_back(it->getMetaValue("IonMatchError"));
+            intensities.push_back(it->getIntensity());
+            match_intensity += it->getIntensity();
+            mzs.push_back(it->getMZ());
+            String ion_name = it->getMetaValue("IonName");
+            {
+              ions.push_back(ion_name);
+              if (terminal_series_match_ratio_)
+              {
+                if (boost::regex_match(ion_name, nt_regex_))
+                {
+                  nint += it->getIntensity();
+                }
+                else if (boost::regex_match(ion_name, ct_regex_))
+                {
+                  cint += it->getIntensity();
+                }
+              }
+              if (max_series_ && boost::regex_match(ion_name, noloss_regex_))
+              {
+                String ion_type = ion_name.prefix(1);
+                if (ListUtils::contains(allowed_types, ion_type)) //case sensitive?
+                {
+                  try
+                  {
+                    int i = ion_name.substr(1).remove('+').toInt() - 1; //do not care about charge
+                    ion_series[ion_type].at(i) = true;
```

Comment:
Yeah, there is a lot of interesting stuff to look at. I'll add a TODO from your suggestion